### PR TITLE
Check for python among execution aliases on Windows

### PR DESCRIPTION
**Pull request by viveklak**
Fixes https://github.com/pulumi/pulumi/issues/5388

### CODE REVIEW

File: sdk/python/python.go [hunk 1]
```diff
@@ -51,14 +51,73 @@ func Command(arg ...string) (*exec.Cmd, error) {
...
 	if err != nil {
-		return nil, fmt.Errorf(
-			"Failed to locate any of %q on your PATH.  Have you installed Python 3.6 or greater?",
-			pythonCmds)
+		// second-chance on windows for python being installed through the Windows app store.
+		if runtime.GOOS == windows {
+			pythonPath, err = resolveWindowsExecutionAlias(pythonCmds)
+		}
+		if err != nil {
+			// TODO: Wrap error? The resulting user facing message isn't very pretty though.
```
Comment (justinvp) [inline sdk/python/python.go:59]: I'm fine not wrapping to keep the user facing message nice. Please remove the TODO.

**Review by justinvp — APPROVED** [review]
Aside from removing the TODO comment, LGTM. Were you able to validate the change works on Windows 10?

**Comment by viveklak** [comment]
Failed to update this - sorry. Essentially while the above code gets past the resolution issue, we are unable to launch the windows app store's version of python because Golang's exec implementation performs a `os.Stat` which doesn't seem to support `syscall.FILE_FLAG_OPEN_REPARSE_POINT` - which execution aliases use. `os.Lstat` does support it so I was able to bypass it in my change here but can't find a reasonable way to bypass `os.exec`'s use of `os.Stat`. I have a repro which I need to file with the Go team. Will file and update this PR.

**Comment by lukehoban** [comment]
Is there a way to at least "fail better" for now in this case?  Can we identify when the target file is one of these reparse points and give a more helpful error?

**Comment by justinvp** [comment]
> but can't find a reasonable way to bypass `os.exec`'s use of `os.Stat`

If copying our own implementation of `os.exec` with workaround is unpalatable in the meantime, I'm wondering if we could use some kind of shim as a workaround, e.g. ship an extra batch script on Windows that we can call from `os.exec`, which then calls the app store version of Python we found?

**Comment by lukehoban** [comment]
FWIW - A few more notes from trying this out on a Windows machine.

The following program is a repro of the core issue:

```go
package main

import (
	"fmt"
	"os"
	"os/exec"
)

func do() error {
	res, err := os.Lstat("C:\\Users\\lukeh\\AppData\\Local\\Microsoft\\WindowsApps\\python3.exe")
	if err != nil {
		return err
	}
	fmt.Printf("%v\n", res)
	cmd := exec.Command("C:\\Users\\lukeh\\AppData\\Local\\Microsoft\\WindowsApps\\python3.exe", "-c", "print('hello')")
	btes, err := cmd.CombinedOutput()
	if err != nil {
		return err
	}
	fmt.Printf("%v\n", btes)
	return nil
}

func main() {
	err := do()
	if err != nil {
		panic(err)
	}
}
```

This results in:

```
&{python3.exe 1056 {2654840238 30774702} {2957402258 30850392} {2957402258 30850392} 0 0 2147483675 0 {0 0}  2853655769 458752 49311 false}
panic: exec: "C:\\Users\\lukeh\\AppData\\Local\\Microsoft\\WindowsApps\\python3.exe": file does not exist

goroutine 1 [running]:
main.main()
        C:/Users/lukeh/goexecalias/main.go:27 +0x4b
exit status 2
```

Even though it definitely does exist:

```
PS C:\Users\lukeh\goexecalias> C:\\Users\\lukeh\\AppData\\Local\\Microsoft\\WindowsApps\\python3.exe -c "print('hello')"
hello
```

The `1056` file attribute indicates this is an archive and a reparse point.  The `2147483675` in `Reserved0` apparently indicates this is (see https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-fscc/c8e77b37-3909-4fe6-a4ea-2b9d423b1ee4):

IO_REPARSE_TAG_APPEXECLINK0x8000001B | Used by Universal Windows Platform (UWP) packages to encode information that allows the application to be launched by CreateProcess. Server-side interpretation only, not meaningful over the wire.
-- | --

**Comment by viveklak** [comment]
> Is there a way to at least "fail better" for now in this case? Can we identify when the target file is one of these reparse points and give a more helpful error?

I am definitely leaning towards this option! Thanks for the dive into the flags @lukehoban. I will update this PR to be a guard and look to file the Go bug.

**Comment by joeduffy** [comment]
Can we construct the `*exec.Cmd` struct directly, bypassing the `LookPath` in the `exec.Command` constructor? For instance, `cmd := &exec.Cmd{Path: "manually populated", Args: ...}`. It may require some Windows specific code to resolve the Windows Execution Alias metadata, but heck, we can even just hardcode the `python` case for now since it's the only one causing us troubles right now (until the Go libraries themselves are fixed to handle this).

Is there a bug in the https://github.com/golang/go repo, by the way? I didn't see one.

**Comment by lukehoban** [comment]
> Is there a bug in the https://github.com/golang/go repo, by the way? I didn't see one.

@viveklak was planning to open one.  We have been unable to find anything tracking this issue.

> Can we construct the *exec.Cmd struct directly, bypassing the LookPath in the exec.Command constructor?

Unfortunately no - that's not the one that fails - the problem is ultimately here: https://github.com/golang/go/blob/master/src/os/exec/exec.go#L352 (and we could possibly avoid that one, but it doesn't look like we can avoid at least one of the `LookPath` calls there).  AFAICT - there is no way to use the existing `exec` package without hitting this.  And it looks like we would have to port over a copy of the majority of that package (including a lot of internals) to avoid this without a fix to the packge.

**Comment by lukehoban** [comment]
Actually - it does look like we could use ` os.StartProcess` directly instead of the `exec` package to workaround this.

**Comment by viveklak** [comment]
I cut https://github.com/golang/go/issues/42919 with the repro and details. Investigating some of the suggestions above next.

**Comment by viveklak** [comment]
> Actually - it does look like we could use ` os.StartProcess` directly instead of the `exec` package to workaround this.

Confirmed that `os.RunCommand` does indeed work at least as far as executing the interpreter. I will try and fit it in tomorrow and update this PR accordingly.

**Comment by viveklak** [comment]
@justinvp @lukehoban I wrapped the `exec.Cmd` usage in an adapter that can fall back to `os.StartProcess` in the specific situation affected by https://github.com/golang/go/issues/42919. This turned into a bit of a yak shave but the changes in https://github.com/pulumi/pulumi/pull/5775/commits/47525eabfa882e42fea673e376acff25b7f9c4a1 are as minimal as I could get it to make `os.StartProcess` actually work for our various use-cases.

I tested this on a windows machine and could confirm that it works.

File: sdk/python/python.go [hunk 2]
```diff
@@ -27,9 +27,51 @@ import (
...
+		cmd.cmd.Stdout = stdout
+	})
+}
+
+func WithStdErr(stderr *os.File) CommandOpt {
+	return commandOptFunc(func(cmd *wrappedCmd) {
+		cmd.stderr = stderr
+		cmd.cmd.Stderr = stderr
+	})
+}
```
Comment (lukehoban) [inline sdk/python/python.go:70]: I didn't exactly follow why we needed to add `CommandOpt` here - vs. just relying on string varargs as before?

File: sdk/python/wrappedcmd.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+	"io"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"syscall"
+)
+
+// WrappedCmd wraps the key portions of exec.Cmd used
+// to launch the python interpreter. Large portions of this
```
[the remark this answers is not in the document]
Comment (lukehoban) [reply sdk/python/wrappedcmd.go:14]: Also - perhaps a `TODO` to remove this all once https://github.com/golang/go/issues/42919 is fixed just so we record that this is all an ugly workaround for that bug?

**Review by lukehoban — APPROVED** [review]
Would love @justinvp's review as well - but LGTM.

**Review by justinvp — APPROVED** [review]
LGTM

**Comment by justinvp** [comment]
Ah, sorry, @viveklak, my other change introduced some conflicts here. Let me know if you have any questions.

**Comment by viveklak** [comment]
Closing in favor of #5874.

**Pull request closed without merging**